app/frontend/wallet/account-manager.ts

```diff
@@ -46,29 +48,68 @@ const AccountManager = ({
   }
 
   async function discoverAccounts() {
+    if (
+      maxAccountIndex === accounts.length - 1 ||
+      (accounts.length > 0 && !(await accounts[accounts.length - 1].isAccountUsed()))
+    ) {
+      return accounts
+    }
+
     const isBulkExportSupported = cryptoProvider.isFeatureSupported(
       CryptoProviderFeature.BULK_EXPORT
     )
-    const shouldExplore =
-      config.shouldExportPubKeyBulk && config.isShelleyCompatible && isBulkExportSupported
-    async function _discoverNextAccount(accountIndex: number) {
-      const newAccount = accounts[accountIndex] || discoverNextAccount()
-      const isAccountUsed = await newAccount.isAccountUsed()
-      if (accountIndex === accounts.length) await addNextAccount(newAccount)
-
-      return (
-        shouldExplore &&
-        isAccountUsed &&
-        accountIndex < maxAccountIndex &&
-        (await _discoverNextAccount(accountIndex + 1))
+
+    const shouldExploreOnlyOne = !(
+      config.shouldExportPubKeyBulk &&
+      config.isShelleyCompatible &&
+      isBulkExportSupported
+    )
+
+    if (shouldExploreOnlyOne) {
+      await exploreNextAccount()
+      return accounts
+    }
+
+    for (const [accountIndexStart, accountIndexEnd] of makeBulkAccountIndexIterator()) {
+      const _accountIndexStart = Math.max(accountIndexStart, accounts.length)
+      const _accountIndexEnd = Math.min(accountIndexEnd, maxAccountIndex)
+
+      if (_accountIndexStart > _accountIndexEnd) {
+        continue
+      }
+
+      const newAccountBatch = _.range(_accountIndexStart, _accountIndexEnd + 1).map(
+        (accountIndex) => accounts[accountIndex] || discoverAccount(accountIndex)
       )
+
+      // Needs to be sequential because HW wallets can't process parallel calls
+      // xpubs keys are used in the next step when deciding if accounts are used
+      for (const account of newAccountBatch) {
+        await account.ensureXpubIsExported()
+      }
+
+      const firstUnusedNewAccountIndex = (
+        await Promise.all(newAccountBatch.map((newAccount) => newAccount.isAccountUsed()))
+      ).findIndex((isUsed) => isUsed === false)
+
+      const foundUnusedAccount = firstUnusedNewAccountIndex !== -1
+
+      const accountsToAdd = newAccountBatch.slice(
+        0,
+        !foundUnusedAccount ? newAccountBatch.length : firstUnusedNewAccountIndex + 1
```

Comment:
nitpick, I'd find it more natural to reverse the condition
```suggestion
        foundUnusedAccount ? firstUnusedNewAccountIndex + 1 : newAccountBatch.length
```